src/elfpy/markets.py

```diff
@@ -67,67 +95,117 @@ def __init__(
+            # pricing model computes new market deltas
+            # market updates its "liquidity pool" wallet, which stores each trade's mint time and user address
+            # LP tokens are also storeds in user wallet as fungible amounts, for ease of user
+            market_deltas, agent_deltas = self._add_liquidity(agent_action)
+            pass
+        elif agent_action.action_type == "remove_liquidity":
+            # market figures out how much the user has contributed (calcualtes their fee weighting)
+            # market resolves fees, adds this to the agent_action
+            # pricing model computes new market deltas
+            # market updates its "liquidity pool" wallet, which stores each trade's mint time and user address
+            # LP tokens are also storeds in user wallet as fungible amounts, for ease of user
+            # TODO: implement fee attribution and withdrawal
+            market_deltas, agent_deltas = self._remove_liquidity(agent_action)
+            pass
         else:
-            raise ValueError(f'ERROR: Unknown trade type "{user_action["action_type"]}".')
+            raise ValueError(f'ERROR: Unknown trade type "{agent_action["action_type"]}".')
         # update market state
         self.update_market(market_deltas)
-        # TODO: self.update_LP_pool(wallet_deltas["fees"])
-        return wallet_deltas
-
+        self.update_spot_price_and_rate()
+        print(f"{self.get_market_step_string()}")
+        # TODO: self.update_LP_pool(agent_deltas["fees"])
+        agent_action.agent.update_wallet(agent_deltas)  # update user state since market doesn't know about users
+        if self.verbose:
+            print(f" agent  Δs ={agent_deltas.display()}\n market Δs ={market_deltas.display()}")
+            print(f" post-trade {agent_action.agent.status_report()}") if self.verbose else None
+    
     def update_market(self, market_deltas):
         """
         Increments member variables to reflect current market conditions
         """
-        for key, value in market_deltas.items():
-            assert np.isfinite(value), f"markets.update_market: ERROR: market delta key {key} is not finite."
-        self.share_reserves += market_deltas["d_base_asset"]
-        self.bond_reserves += market_deltas["d_token_asset"]
-        self.cum_base_asset_slippage += market_deltas["d_base_asset_slippage"]
-        self.cum_token_asset_slippage += market_deltas["d_token_asset_slippage"]
-        self.cum_base_asset_fees += market_deltas["d_base_asset_fee"]
-        self.cum_token_asset_fees += market_deltas["d_token_asset_fee"]
-        self.base_asset_orders += market_deltas["d_base_asset_orders"]
-        self.token_asset_orders += market_deltas["d_token_asset_orders"]
-        self.base_asset_volume += market_deltas["d_base_asset_volume"]
-        self.token_asset_volume += market_deltas["d_token_asset_volume"]
+        for field in market_deltas.__dataclass_fields__:
```

Comment:
isn't this resolved more easily with the `BasicDataclass` additions? like `market_deltas.items()` or whatever?